adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1395 @@
+	if bidderRawResponse.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	httpStatusError := checkRespStatusCode(bidderRawResponse)
+	if httpStatusError != nil {
+		return nil, []error{httpStatusError}
+	}
+
+	var huaweiAdsResponse HuaweiAdsResponse
+	if err := json.Unmarshal(bidderRawResponse.Body, &huaweiAdsResponse); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "Bad Server Response",
+		}}
+	}
+
+	huaweiAdsResponseError := checkHuaweiAdsResponseRetcode(huaweiAdsResponse)
+	if huaweiAdsResponseError != nil {
+		return nil, []error{huaweiAdsResponseError}
+	}
+
+	var err error
+	bidderResponse, err = a.convertHuaweiAdsResp2BidderResp(&huaweiAdsResponse, openRTBRequest)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return bidderResponse, nil
+}
+
+// Builder builds a new instance of the HuaweiAds adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	extraInfo, err := getExtraInfo(config.ExtraAdapterInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	bidder := &adapter{
+		endpoint:  config.Endpoint,
+		extraInfo: extraInfo,
+	}
+	return bidder, nil
+}
+
+func getExtraInfo(huaweiAdsExtraInfo string) (ExtraInfo, error) {
+	if huaweiAdsExtraInfo == "" {
+		return ExtraInfo{
+			TrackingUrl: DefaultTrackingEndpoint,
+		}, nil
+	}
+
+	var extraInfo ExtraInfo
+	if err := json.Unmarshal([]byte(huaweiAdsExtraInfo), &extraInfo); err != nil {
+		return extraInfo, fmt.Errorf("HuaweiAds ExtraInfo: invalid extra info: %v", err)
+	}
+
+	if extraInfo.TrackingUrl == "" {
+		extraInfo.TrackingUrl = DefaultTrackingEndpoint
+	}
+	return extraInfo, nil
+}
+
+// get request header
+func getHeaders(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, request *openrtb2.BidRequest, isAddAuthorization bool) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	if huaweiAdsImpExt == nil {
+		return headers
+	}
+
+	if isAddAuthorization {
+		headers.Add("Authorization", getDigestAuthorization(huaweiAdsImpExt))
+	}
+
+	if request.Device != nil && len(request.Device.UA) > 0 {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+	return headers
+}
+
+// get body json for HuaweiAds request
+func getHuaweiAdsReqJson(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest, huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) error {
+	request.Version = HuaweiAdxApiVersion
+	var err error
+	if err = getHuaweiAdsReqAppInfo(request, openRTBRequest); err != nil {
+		return err
+	}
+	if err = getHuaweiAdsReqDeviceInfo(request, openRTBRequest, huaweiAdsImpExt); err != nil {
+		return err
+	}
+	getHuaweiAdsReqNetWorkInfo(request, openRTBRequest)
+	getHuaweiAdsReqRegsInfo(request, openRTBRequest)
+	getHuaweiAdsReqGeoInfo(request, openRTBRequest)
+	return nil
+}
+
+// get adslot30
```

Comment:
You should be able to drop this comment, or in the alternative describe what this function is supposed to do.

Reply:
deleted